diff --git a/pkg/caption/caption.go b/pkg/caption/caption.go
index 69b3ed3a..4e6358dd 100644
--- a/pkg/caption/caption.go
+++ b/pkg/caption/caption.go
@@ -269,7 +269,7 @@ func (c *Caption) Download(writer io.Writer) error {
 		return errors.Join(errDownloadCaption, err)
 	}
 
-	file, err := os.Create(c.File)
+	file, err := pkg.Root.OpenFile(c.File, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
 	if err != nil {
 		return errors.Join(errDownloadCaption, err)
 	}
