diff --git a/cng/tls1prf.go b/cng/tls1prf.go
index 5a3fb01..56131a6 100644
--- a/cng/tls1prf.go
+++ b/cng/tls1prf.go
@@ -44,6 +44,7 @@ func TLS1PRF(result, secret, label, seed []byte, h func() hash.Hash) error {
 	if err := bcrypt.GenerateSymmetricKey(alg, &kh, nil, secret, 0); err != nil {
 		return err
 	}
+	defer bcrypt.DestroyKey(kh)
 
 	buffers := make([]bcrypt.Buffer, 0, 3)
 	if len(label) > 0 {
