From 489af403a19057f6b6b4b1dc0e48cbb26a202ef9 Mon Sep 17 00:00:00 2001
From: Henrique Dias <mail@hacdias.com>
Date: Sun, 8 Feb 2026 08:02:03 +0100
Subject: [PATCH] fix: remove skip clean

---
 http/http.go | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/http/http.go b/http/http.go
index ec8e52c156f..0ddca9692cf 100644
--- a/http/http.go
+++ b/http/http.go
@@ -35,11 +35,6 @@ func NewHandler(
 	})
 	index, static := getStaticHandlers(store, server, assetsFs)
 
-	// NOTE: This fixes the issue where it would redirect if people did not put a
-	// trailing slash in the end. I hate this decision since this allows some awful
-	// URLs https://www.gorillatoolkit.org/pkg/mux#Router.SkipClean
-	r = r.SkipClean(true)
-
 	monkey := func(fn handleFunc, prefix string) http.Handler {
 		return handle(fn, prefix, store, server)
 	}
