From 3ce9f41c54d04a65f10b23a46771391c00659afb Mon Sep 17 00:00:00 2001
From: Luboslav Pivarc <lpivarc@redhat.com>
Date: Tue, 22 Jul 2025 11:59:30 +0200
Subject: [PATCH] Detector: Use safepath

Signed-off-by: Luboslav Pivarc <lpivarc@redhat.com>
---
 pkg/virt-handler/isolation/detector.go | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/pkg/virt-handler/isolation/detector.go b/pkg/virt-handler/isolation/detector.go
index 5f636fa8a4a1..0e2c62567b7b 100644
--- a/pkg/virt-handler/isolation/detector.go
+++ b/pkg/virt-handler/isolation/detector.go
@@ -36,6 +36,7 @@ import (
 	v1 "kubevirt.io/api/core/v1"
 	"kubevirt.io/client-go/log"
 
+	"kubevirt.io/kubevirt/pkg/safepath"
 	"kubevirt.io/kubevirt/pkg/util"
 	"kubevirt.io/kubevirt/pkg/virt-controller/services"
 	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
@@ -214,6 +215,10 @@ func (s *socketBasedIsolationDetector) getPid(socket string) (int, error) {
 	}
 	defer sock.Close()
 
+	_, err = safepath.NewPathNoFollow(socket)
+	if err != nil {
+		return -1, err
+	}
 	ufile, err := sock.(*net.UnixConn).File()
 	if err != nil {
 		return -1, err
