From c1fa2b13f6b547b96da60b23350bbe2b29de542d Mon Sep 17 00:00:00 2001
From: aichy126 <16996097+aichy126@users.noreply.github.com>
Date: Tue, 31 Jan 2023 10:04:52 +0800
Subject: [PATCH] fix  RetrievePassWord code

---
 internal/controller/user_controller.go | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/internal/controller/user_controller.go b/internal/controller/user_controller.go
index 909581ccf..c2cf25bfa 100644
--- a/internal/controller/user_controller.go
+++ b/internal/controller/user_controller.go
@@ -157,8 +157,8 @@ func (uc *UserController) RetrievePassWord(ctx *gin.Context) {
 		return
 	}
 	_, _ = uc.actionService.ActionRecordAdd(ctx, schema.ActionRecordTypeFindPass, ctx.ClientIP())
-	code, err := uc.userService.RetrievePassWord(ctx, req)
-	handler.HandleResponse(ctx, err, code)
+	_, err := uc.userService.RetrievePassWord(ctx, req)
+	handler.HandleResponse(ctx, err, nil)
 }
 
 // UseRePassWord godoc
